### PR TITLE
[mlir][sparse] fix stack overflow due to alloca instruction inside loops

### PR DESCRIPTION
**Pull request by PeimingLiu**

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp [hunk 1]
```diff
@@ -589,14 +589,23 @@ class SparseTensorInsertConverter : public OpConversionPattern<InsertOp> {
     const auto stt = getSparseTensorType(op.getTensor());
     const auto elemTp = stt.getElementType();
     const Level lvlRank = stt.getLvlRank();
+    // We need a alloc scope here as the InsertOp is always generated inside a
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp:592]: an alloc scope

**Review by aartbik — APPROVED** [review]
Looks right to me, but please wait for Alex to chime in as well.

File: mlir/include/mlir/Dialect/SCF/IR/SCFOps.td [hunk 2]
```diff
@@ -120,7 +120,7 @@ def ExecuteRegionOp : SCF_Op<"execute_region", [
 //===----------------------------------------------------------------------===//
 
 def ForOp : SCF_Op<"for",
-      [AutomaticAllocationScope, DeclareOpInterfaceMethods<LoopLikeOpInterface,
```
Comment (PeimingLiu) [inline mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:123]: I don't think `scf.forall` and `scf.parallel_for` should have the trait either.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp [hunk 3]
```diff
@@ -589,14 +589,23 @@ class SparseTensorInsertConverter : public OpConversionPattern<InsertOp> {
     const auto stt = getSparseTensorType(op.getTensor());
     const auto elemTp = stt.getElementType();
     const Level lvlRank = stt.getLvlRank();
+    // We need an alloca scope here as the InsertOp is always generated inside a
+    // loop, which lead to stack overflow.
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp:593]: leads?

**Review by yinying-lisa-li — APPROVED** [review]

**Comment by ftynse** [comment]
Ah, the infamous allocation-in-the-loop problem. I do _not_ think we should remove the trait from `scf.for`. The fact that the lowering doesn't do stack save/restore doesn't mean that the semantics of the op is different. Semantics are not defined by lowerings, and the lowering may be incorrect. I remember we had very long and inconclusive discussions on whether this lowering should insert stack save/restore. Even if we did remove the trait, it absolutely should not happen in a change tagged "mlir, sparse, fix <...>".

**Review by ftynse — CHANGES_REQUESTED** [review]
we shouldn't just drop the trait here.

**Comment by PeimingLiu** [comment]
> do stack save/restore doesn't mean that the semantics of the op is different. Semantics are not defined by lowerings, and the lowering may be incorrect. I remember we had very long and inconclusive discussions on whether this lowering should insert stack save/restore. Even if we did remove the trait, it absolutely should not happen in a change tagged "mlir, sparse, fix <...>".

Thanks for the insights, the reason why I need to remove the trait here is due to `AllocaScopeHoister`, which hoist `memref::alloca` from the `alloca_scope` into `scf.for`, leading to incorrect code.

E.g.,
```
scf.for {
  memref.alloca_scope {
     memref.alloca
     xxx
  }
}
```
is canonicalized into 
```
scf.for {
  memref.alloca
  memref.alloca_scope {
    xxx
  }
}
```

Note that this is a correct transformation **provided that `scf.for` indeed adjusts stack pointer**. Since it now does not, what is your suggestion? I think insert `llvm.stack.save` and `llvm.stack.restore` should also works.

I can of course submit the scf.for related change in a separate PR to address your second concern;-)

**Comment by ftynse** [comment]
Let's bring this to discourse, I remember there were several folks who had different views on this problem.

**Comment by PeimingLiu** [comment]
https://discourse.llvm.org/t/should-scf-for-has-automaticallocationscope-trait/74281

**Comment by PeimingLiu** [comment]
Also, we worked around the issue in `scf.for` by https://github.com/llvm/llvm-project/pull/69786. (though I still think the concern need to be addressed)

**Pull request closed without merging**